pkg/util/testutil/docker/run.go

```diff
@@ -0,0 +1,138 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build test
+
+package docker
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"testing"
+	"time"
+)
+
+func Run(t testing.TB, cfg LifecycleConfig) error {
+	var err error
+	var ctx context.Context
+	for i := 0; i < cfg.Retries(); i++ {
+		t.Helper()
+		// Ensuring no previous instances exists.
+		killPreviousInstances(cfg)
+
+		scanner := NewScanner(cfg.LogPattern(), make(chan struct{}, 1))
+		// attempt to start the container/s
+		ctx, err = run(t, cfg, scanner)
+		if err != nil {
+			t.Logf("could not start %s: %v", cfg.Name(), err)
+			//this iteration failed, retry
+			continue
+		}
+
+		//check container logs for successful start
+		if err = checkReadiness(ctx, cfg, scanner); err == nil {
+			// target container/s started successfully, we can stop the retries loop and finish here
+			return nil
+		}
+		scanner.PrintLogs(t)
+		t.Logf("failed to start %s server, retrying: %v", cfg.Name(), err)
+		time.Sleep(5 * time.Second)
+	}
+	return err
+}
+
+// isStart is true if command is to start the container, false if to stop the container
+func buildCommandArgs(cfg LifecycleConfig, isStart bool) []string {
+	var args []string
+	switch concreteCfg := cfg.(type) {
+	case runConfig:
+		if isStart {
```

Comment:
shouldn't this be a method of the interface, instead of a switch based on the type?